scratchpad/issue-258.md

```diff
@@ -1,34 +1,78 @@
+    - If `auto_mount` is true, call `ensure_mounted()`.
+    - If root does not exist and `auto_mount` is false, return `Error::NotMounted`.
+    - List immediate children of the configured root; filter to entries that are directories AND whose name starts with `self.cfg.group_prefix`.
+    - For each candidate, call `FsProvider::remove_dir` and increment a counter on success. On `ENOENT`, skip (idempotent). On other errors, let the library continue (best-effort) and bubble a summary only if the root listing failed at the start.
+    - Return `Ok(removed_count)`.
+  - Keep sanitization minimal: only strict prefix match; do not attempt to parse UIDs here. Rely on group creation to sanitize UIDs.
 
-## Implementation Notes
-- Use strict prefix match and simple sanitization to avoid accidental deletion.
-- Consider a dry-run option or debug logging of candidates.
+- nri-resctrl-plugin
+  - Add an internal `did_startup_cleanup: AtomicBool` (or `OnceCell`) to `ResctrlPlugin`.
+  - On first `synchronize()` when `cfg.cleanup_on_start` is true, call `resctrl.cleanup_all()` and set the flag. If it returns `Error::NotMounted`, log at debug and proceed; otherwise log removed count at info/debug. Never emit `PodResctrlEvent` for cleanup-only actions.
+  - Unit-test this behavior using the plugin’s test FS and DI constructor (`with_resctrl` / `with_pid_source`).
 
-## Dependencies
-- Sub-Issue 01 (crate), Sub-Issue 03 (plugin skeleton).
+## Detailed Test Plan
 
-## Testing
-- Mock FS directory listing of resctrl root; ensure only expected entries removed.
+- resctrl crate unit tests (using mock `FsProvider`):
+  - Root not mounted, `auto_mount=false`: `cleanup_all()` returns `Error::NotMounted` and removes nothing.
+  - Root not mounted, `auto_mount=true`: mock mount succeeds; `cleanup_all()` proceeds and removes matching dirs.
+  - Mixed entries: under `/sys/fs/resctrl`, create entries `pod_a`, `pod_b`, `info` (dir), `tasks` (file), `other` (dir). Only `pod_*` are removed; others remain. Assert returned count is 2.
+  - Permission denied on one directory: `remove_dir` for `pod_b` returns `EACCES`; method still returns `Ok(1)` (best-effort) and other entries remain untouched.
+  - Idempotency: calling `cleanup_all()` twice removes the same initial set once; second call returns 0 and is not an error.
 
-## Risks
-- Accidental deletion mitigated by strict prefix filter and tests.
+- nri-resctrl-plugin tests:
+  - Startup once: pre-populate resctrl root with `pod_stale1`, `pod_stale2`. Create plugin with `cleanup_on_start=true`. Call `synchronize()` with empty state. Assert both directories are removed and no events were received on the channel. Call `synchronize()` again; assert no additional deletions occur (flag prevents re-run).
+  - Not mounted: when root missing and `auto_mount=false`, `cleanup_all()` returns `NotMounted`; plugin logs but continues. Verify no panic, no events, and directories obviously remain absent.
```

Comment:
since it is the responsibility of the plugin to call ensure_mounted(), this checks functionality outside this spec. But, please check if there is no test for the auto_mount=false case, then we should add it as part of this spec (but we should be clear that we found a test was missing for the auto-mount functionality, not part of this work)